File: src/main/java/org/moinex/ui/dialog/AddCategoryController.java
/*
 * Filename: AddCategoryController.java
 * Created on: October 13, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog;

import jakarta.persistence.EntityExistsException;
import java.text.MessageFormat;
import javafx.fxml.FXML;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.Category;
import org.moinex.service.CategoryService;
import org.moinex.service.PreferencesService;
import org.moinex.util.Constants;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

/** Controller for the Add Category dialog */
@Controller
@NoArgsConstructor
public class AddCategoryController {
    @FXML private TextField categoryNameField;

    private CategoryService categoryService;

    private PreferencesService preferencesService;

    @Autowired
    public AddCategoryController(
            CategoryService categoryService, PreferencesService preferencesService) {
        this.categoryService = categoryService;
        this.preferencesService = preferencesService;
    }

    @FXML
    private void initialize() {
        // Still empty
    }

    @FXML
    public void handleSave() {
        String name = categoryNameField.getText();

        try {
            categoryService.createCategory(new Category(null, name, false));

            WindowUtils.showSuccessDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.CATEGORY_DIALOG_CATEGORY_ADDED_TITLE),
                    MessageFormat.format(
                            preferencesService.translate(
                                    Constants.TranslationKeys
                                            .CATEGORY_DIALOG_CATEGORY_ADDED_MESSAGE),
                            name));

            Stage stage = (Stage) categoryNameField.getScene().getWindow();
            stage.close();
        } catch (IllegalArgumentException | EntityExistsException e) {
            WindowUtils.showErrorDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.CATEGORY_DIALOG_ERROR_ADDING_CATEGORY_TITLE),
                    e.getMessage());
        }
    }

    @FXML
    public void handleCancel() {
        Stage stage = (Stage) categoryNameField.getScene().getWindow();
        stage.close();
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/AddCategoryController.kt
/*
 * Filename: AddCategoryController.kt (original filename: AddCategoryController.java)
 * Created on: October 13, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog

import javafx.fxml.FXML
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.WindowUtils
import org.moinex.model.Category
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.springframework.stereotype.Controller

@Controller
class AddCategoryController(
    private val categoryService: CategoryService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var categoryNameField: TextField

    @FXML
    private fun initialize() {
        // Still empty
    }

    @FXML
    private fun handleSave() {
        val name = categoryNameField.text

        runCatching {
            categoryService.createCategory(Category(name = name, isArchived = false))

            (categoryNameField.scene.window as Stage).close()
        }.onFailure { e ->
            WindowUtils.showErrorDialog(
                preferencesService.translate(TranslationKeys.CATEGORY_DIALOG_ERROR_ADDING_CATEGORY_TITLE),
                e.message ?: "Unknown error",
            )
        }
    }

    @FXML
    private fun handleCancel() {
        (categoryNameField.scene.window as Stage).close()
    }
}
